packages/core/src/publicCredential/PublicCredential.chain.ts

```diff
@@ -0,0 +1,230 @@
+import type {
+  PublicCredentialError,
+  PublicCredentialsCredentialsCredential,
+  PublicCredentialsCredentialsCredentialEntry,
+} from '@kiltprotocol/augment-api'
+
+import { encode as cborEncode, decode as cborDecode } from 'cbor'
+
+import { hexToU8a, u8aToHex } from '@polkadot/util'
+import { HexString } from '@polkadot/util/types'
+import { ConfigService } from '@kiltprotocol/config'
+import { fromChain as didFromChain } from '@kiltprotocol/did'
+import { validateUri } from '@kiltprotocol/asset-did'
+import { SDKErrors } from '@kiltprotocol/utils'
+
+import { computeId } from './PublicCredential.js'
+
+export interface EncodedPublicCredential {
+  ctypeHash: CTypeHash
+  subject: AssetDidUri
+  claims: HexString
+  authorization: IDelegationNode['id'] | null
+}
+
+/**
+ * Format a [[IPublicCredentialInput]] to be used as a parameter for the blockchain API function.
+
+ * @param publicCredential The public credential to format.
+ * @returns The blockchain-formatted public credential.
+ */
+export function toChain(
+  publicCredential: IPublicCredentialInput
+): EncodedPublicCredential {
+  const { cTypeHash, claims, subject, delegationId } = publicCredential
+
+  const cborSerializedClaims = cborEncode(claims)
+
+  return {
+    ctypeHash: cTypeHash,
+    subject,
+    // FIXME: Using Uint8Array directly fails to encode and decode, I guess because the api object assumes the byte array is SCALE-encoded.
+    claims: u8aToHex(new Uint8Array(cborSerializedClaims)),
+    authorization: delegationId,
+  }
+}
+
+// Flatten any nested batch calls into a single list of calls.
+function flattenCalls(api: ApiPromise, call: Call): Call[] {
+  if (
+    api.tx.utility.batch.is(call) ||
+    api.tx.utility.batchAll.is(call) ||
+    api.tx.utility.forceBatch.is(call)
+  ) {
+    // Inductive case
+    return call.args[0].reduce(
+      (acc: Call[], c: Call) => acc.concat(flattenCalls(api, c)),
+      []
+    )
+  }
+  // Base case
+  return [call]
+}
+
+// Transform a blockchain-formatted public credential into the original [[IPublicCredentialInput]].
+// It throws if what was written on the chain was garbage.
+function credentialInputFromChain(
+  credential: PublicCredentialsCredentialsCredential
+): IPublicCredentialInput {
+  validateUri(credential.subject.toUtf8())
+  return {
+    claims: cborDecode(hexToU8a(credential.claims.toHex())),
+    cTypeHash: credential.ctypeHash.toHex(),
+    delegationId: credential.authorization.unwrapOr(undefined)?.toHex() ?? null,
+    subject: credential.subject.toUtf8() as AssetDidUri,
+  }
+}
+
+// Retrieve a given block and looks into it to find a public credential creation tx that matches the provided credential ID.
+async function retrievePublicCredentialCreationExtrinsicsFromBlock(
+  api: ApiPromise,
+  credentialId: HexString,
+  blockNumber: u64
+): Promise<Extrinsic | null> {
+  const { extrinsics } = await api.derive.chain.getBlockByNumber(blockNumber)
```

Comment:
Should they be flattened as well?

Reply:
What do you mean?

Reply:
I didn’t notice that below you look at events, not calls, so it probably doesn’t matter that some calls could have been batch calls.